pkg/tsdb/loki/frame.go

```diff
@@ -51,9 +52,31 @@ func adjustFrame(frame *data.Frame, query *lokiQuery) *data.Frame {
 		field.Config.DisplayNameFromDS = name
 	}
 
+	// for streams-dataframes, we need to send to the browser the nanosecond-precision timestamp too.
```

Comment:
So it's for stream frames that we are adding this new tsNs field? Because it looks like we have it for [non-streaming](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L567) frames.

A second question... I see we are testing the streaming frame by running the query on explore? But it's running through `api/ds/query`. I thought streaming frames were only used when running live tail mode in explore?

Reply:
sorry, misunderstanding. the "stream" word here does not mean the live-tailing live-streaming thing. it's how Loki calls data-response-containing-logs (not numbers).

for example, check this: https://grafana.com/docs/loki/latest/api/#get-lokiapiv1query_range

it can return a "matrix values" or a "stream values"...